src/de.rs

```diff
@@ -0,0 +1,491 @@
+use pyo3::type_object::PyTypeInfo;
+use pyo3::types::*;
+use pyo3::{PyObject, Python};
+use serde::{de, Deserialize};
+
+use crate::error::{PythonizeError, Result};
+
+pub fn depythonize<'de, T>(py: Python, obj: PyObject) -> Result<T>
+where
+    T: Deserialize<'de>,
+{
+    let mut depythonizer = Depythonizer::from_object(py, obj);
+    T::deserialize(&mut depythonizer)
+}
+
+#[derive(Debug)]
+enum GetItemKey {
+    Key(PyObject),
+    Index(isize),
+    None,
+}
+
+pub struct Depythonizer<'py> {
+    input: PyObject,
+    current: GetItemKey,
+    py: Python<'py>,
+}
```

Comment:
Similarly can store `&'py PyAny` (which implicitly holds the GIL).

```suggestion
pub struct Depythonizer<'py> {
    input: &'py PyAny,
    current: GetItemKey,
}
```

Reply:
Implemented, this does make it simpler, and I wasn't initially certain of the cost for creating a new PyObject